includes/RemovePIIJob.php

```diff
@@ -0,0 +1,489 @@
+			]
+		);
+
+		$user = User::newSystemUser( 'MediaWiki default', [ 'steal' => true ] );
+
+		if ( !$user ) {
+			$this->setLastError( 'Invalid username' );
+
+			return false;
+		}
+
+		$userGroupManager = MediaWikiServices::getInstance()->getUserGroupManager();
+
+		// Hide deletions from RecentChanges
+		$userGroupManager->addUserToGroup( $user, 'bot', null, true );
+
+		$dbr = wfGetDB( DB_REPLICA, [], $this->database );
+		$userPageTitle = $oldName->getUserPage();
+
+		$namespaces = [
+			NS_USER,
+			NS_USER_TALK
+		];
+		
+		if ( class_exists( 'UserProfilePage' ) ) {
+			array_push( $namespaces,
+				NS_USER_WIKI,
+				NS_USER_WIKI_TALK,
+				NS_USER_PROFILE,
+				NS_USER_PROFILE_TALK
+			);
+		}
+		
+		if ( ExtensionRegistry::getInstance()->isLoaded( 'BlogPage' ) ) {
+			array_push( $namespaces,
+				NS_BLOG,
+				NS_BLOG_TALK
+			);
+		}
+
+		if ( ExtensionRegistry::getInstance()->isLoaded( 'SimpleBlogPage' ) ) {
+			array_push( $namespaces,
+				NS_USER_BLOG,
+				NS_USER_BLOG_TALK
+			);
+		}
+
+		$rows = $dbr->select(
+			'page', [
+				'page_namespace',
+				'page_title'
+			], [
+				'page_namespace IN (' . implode( ',', $namespaces ) . ')',
+				'(page_title ' . $dbr->buildLike( $userPageTitle->getDBkey() . '/', $dbr->anyString() ) .
+				' OR page_title = ' . $dbr->addQuotes( $userPageTitle->getDBkey() ) . ')'
+			],
+			__METHOD__
+		);
+
+		$dbw->query( 'DELETE FROM revision WHERE rev_id IN (SELECT rev_id FROM `revision` LEFT JOIN `page` ON rev_page = page_id WHERE' . '(page_title ' . $dbw->buildLike( $userPageTitle->getDBkey() . '/', $dbw->anyString() ) .
+				' OR page_title = ' . $dbw->addQuotes( $userPageTitle->getDBkey() ) . '))' );
+
+		$error = '';
+		foreach ( $rows as $row ) {
+			$title = Title::newFromRow( $row );
+
+			$userPage = WikiPage::factory( $title );
+
+			AtEase::suppressWarnings();
+
+			$status = $userPage->doDeleteArticleReal( '', $user, true, null, $error, null, [], 'delete', true );
+
+			AtEase::restoreWarnings();
+
+			if ( !$status->isOK() ) {
+				$errorMessage = json_encode( $status->getErrorsByType( 'error' ) );
+				$this->setLastError( "Failed to delete user {$userOldName} page, likely does not have a user page. Error: {$errorMessage}" );
+
+				continue;
+			}
+		}
+		
+		$dbw->delete(
+			'logging', [
+				'(log_title ' . $dbr->buildLike( $userPageTitle->getDBkey() . '/', $dbr->anyString() ) .
+				' OR log_title = ' . $dbr->addQuotes( $userPageTitle->getDBkey() ) . ')'
+			]
+		);
+
+		$dbw->delete(
+			'recentchanges', [
+				'(rc_title ' . $dbr->buildLike( $userPageTitle->getDBkey() . '/', $dbr->anyString() ) .
+				' OR rc_title = ' . $dbr->addQuotes( $userPageTitle->getDBkey() ) . ')'
+			]
+		);
+
+		// Lock global account
+		$newCentral->adminLock();
+
+		// Invalidate cache now
+		$newCentral->invalidateCache();
+
+		// Remove user email
+		if ( $newName->getEmail() ) {
+			$newName->invalidateEmail();
+			$newName->saveSettings();
+		}
+
+		return true;
+	}
```

Comment:
If we can make the rename never log then we can likely combine this all into one step. Adapt the above code to be the first thing to run, once all that's done automatically start the rename job. This would definitely simplify the whole thing and make the old username never shown in any logs.